fea-rs/src/parse/source.rs

```diff
@@ -47,9 +49,108 @@ pub struct SourceMap {
 }
 
 /// An error that occurs when trying to read a file from disk.
+#[derive(Clone, Debug, thiserror::Error)]
+#[error("Failed to load source at '{}': '{cause}'", Path::new(.path.as_os_str()).display())]
 pub struct SourceLoadError {
-    pub(crate) cause: std::io::Error,
-    pub(crate) path: PathBuf,
+    #[source]
+    cause: Rc<dyn std::error::Error>,
+    path: OsString,
+}
+
+/// A trait that abstracts resolving a path.
+///
+/// In general, paths are resolved through the filesystem; however if you are
+/// doing something fancy (such as keeping your source files in memory) you
+/// can pass a closure or another custom implementation of this trait into the
+/// appropriate parse functions.
+///
+/// If you need a custom resolver, you can either implement this trait for some
+/// custom type, or you can use a closure with the signature,
+/// `|&OsStr| -> Result<String, SourceLoadError>`.
```

Comment:
Why OsStr and not Path/PathBuf?

Reply:
It's sort of a philosophical point: since we are abstracting away the idea of these necessarily being paths (in the sense of paths on the filesystem) then I thought it made sense to not use `Path` types to represent them. `OsStr` is more general than `Path`; it can be freely converted back and forth, but doesn't include any of the filesystem or OS specific functionality, such as [particular conventions for path concatenation](http://doc.rust-lang.org/1.66.1/std/path/struct.PathBuf.html#method.push), or methods that involve hitting the filesystem, like [`Path::canonicalize`](http://doc.rust-lang.org/1.66.1/std/path/struct.Path.html#method.canonicalize).